File: sdks/thunkmetrc-py/client/src/thunkmetrc/client/services/sublocations.py
from typing import Any, Optional, Dict, List, Union
import urllib.parse

class SublocationsClient:
    def __init__(self, client):
        self.client = client
    def create_sublocations(self, body: Any = None, license_number: Optional[str] = None) -> Any:
        """
        Creates new sublocation records for a Facility.
        Permissions Required:
        - Manage Locations
        POST CreateSublocations
        Parameters:
        - body (Any): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        query_params = {}
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        path = f"/sublocations/v2"
        return self.client._send("POST", path, body, query_params)
    def delete_sublocations_by_id(self, id: str, license_number: Optional[str] = None) -> Any:
        """
        Archives an existing Sublocation record for a Facility.
        Permissions Required:
        - Manage Locations
        DELETE DeleteSublocationsById
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        query_params = {}
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        path = f"/sublocations/v2/{urllib.parse.quote(id)}"
        return self.client._send("DELETE", path, body, query_params)
    def get_active_sublocations(self, last_modified_end: Optional[str] = None, last_modified_start: Optional[str] = None, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> Any:
        """
        Retrieves a list of active sublocations for the current Facility, optionally filtered by last modified date range.
        Permissions Required:
        - Manage Locations
        GET GetActiveSublocations
        Parameters:
        - last_modified_end (str, optional): Filter by lastModifiedEnd
        - last_modified_start (str, optional): Filter by lastModifiedStart
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        query_params = {}
        if last_modified_end is not None:
            query_params["lastModifiedEnd"] = last_modified_end
        if last_modified_start is not None:
            query_params["lastModifiedStart"] = last_modified_start
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        if page_number is not None:
            query_params["pageNumber"] = page_number
        if page_size is not None:
            query_params["pageSize"] = page_size
        path = f"/sublocations/v2/active"
        return self.client._send("GET", path, body, query_params)
    def get_inactive_sublocations(self, license_number: Optional[str] = None, page_number: Optional[str] = None, page_size: Optional[str] = None) -> Any:
        """
        Retrieves a list of inactive sublocations for the specified Facility.
        Permissions Required:
        - Manage Locations
        GET GetInactiveSublocations
        Parameters:
        - license_number (str, optional): Filter by licenseNumber
        - page_number (str, optional): Filter by pageNumber
        - page_size (str, optional): Filter by pageSize
        """
        query_params = {}
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        if page_number is not None:
            query_params["pageNumber"] = page_number
        if page_size is not None:
            query_params["pageSize"] = page_size
        path = f"/sublocations/v2/inactive"
        return self.client._send("GET", path, body, query_params)
    def get_sublocations_by_id(self, id: str, license_number: Optional[str] = None) -> Any:
        """
        Retrieves a Sublocation by its Id, with an optional license number.
        Permissions Required:
        - Manage Locations
        GET GetSublocationsById
        Parameters:
        - id (str): Path parameter id
        - license_number (str, optional): Filter by licenseNumber
        """
        query_params = {}
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        path = f"/sublocations/v2/{urllib.parse.quote(id)}"
        return self.client._send("GET", path, body, query_params)
    def update_sublocations(self, body: Any = None, license_number: Optional[str] = None) -> Any:
        """
        Updates existing sublocation records for a specified Facility.
        Permissions Required:
        - Manage Locations
        PUT UpdateSublocations
        Parameters:
        - body (Any): Request body
        - license_number (str, optional): Filter by licenseNumber
        """
        query_params = {}
        if license_number is not None:
            query_params["licenseNumber"] = license_number
        path = f"/sublocations/v2"
        return self.client._send("PUT", path, body, query_params)

File: probe/internal/metrc/client/sublocations.go
package client

import (
	"strings"
	"fmt"
	"github.com/thunkier/thunkmetrc/probe/pkg/metrc/models"
)

// CreateSublocations (Sublocations)
// POST {{baseUrl}}/sublocations/v2
//   licenseNumber (required): The License Number of the Facility for which to record the list of sublocations.
func (f *Fetcher) CreateSublocations(body []models.SublocationsRequest, licenseNumber string) (models.WriteResponse, error) {
	url := "{{baseUrl}}/sublocations/v2"
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchOne[models.WriteResponse](f, "Sublocations", "CreateSublocations", "POST", url, body)
}

// DeleteSublocationsById (Sublocations)
// DELETE {{baseUrl}}/sublocations/v2/{id}
//   licenseNumber (required): The License Number of the Facility for which to delete the list of sublocations.
func (f *Fetcher) DeleteSublocationsById(id string, licenseNumber string) (map[string]any, error) {
	url := "{{baseUrl}}/sublocations/v2/{id}"
	url = strings.ReplaceAll(url, "{"+"id"+"}", id)
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchOne[map[string]any](f, "Sublocations", "DeleteSublocationsById", "DELETE", url, nil)
}

// GetActiveSublocations (Sublocations)
// GET {{baseUrl}}/sublocations/v2/active
//   licenseNumber (required): The License Number of the Facility for which to return the list of active sublocations.
//   lastModifiedStart (optional): The Last Modified start timestamp
//   lastModifiedEnd (optional): The Last Modified end timestamp
//   pageNumber (optional): The number of the data page from which to return data.
//   pageSize (optional): The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.
func (f *Fetcher) GetActiveSublocations(licenseNumber string, lastModifiedStart string, lastModifiedEnd string, pageNumber string, pageSize string) ([]models.Sublocation, error) {
	url := "{{baseUrl}}/sublocations/v2/active"
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}
	if lastModifiedStart != "" {
		queryParams = append(queryParams, fmt.Sprintf("lastModifiedStart=%s", lastModifiedStart))
	}
	if lastModifiedEnd != "" {
		queryParams = append(queryParams, fmt.Sprintf("lastModifiedEnd=%s", lastModifiedEnd))
	}
	if pageNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("pageNumber=%s", pageNumber))
	}
	if pageSize != "" {
		queryParams = append(queryParams, fmt.Sprintf("pageSize=%s", pageSize))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchList[models.Sublocation](f, "Sublocations", "GetActiveSublocations", "GET", url, nil)
}

// GetInactiveSublocations (Sublocations)
// GET {{baseUrl}}/sublocations/v2/inactive
//   licenseNumber (required): The License Number of the Facility for which to return the list of inactive sublocations.
//   pageNumber (optional): The number of the data page from which to return data.
//   pageSize (optional): The number of records to return per page. Pagination is currently disabled by default. You can enable pagination on this query by specifying a value that does not exceed 20.
func (f *Fetcher) GetInactiveSublocations(licenseNumber string, pageNumber string, pageSize string) ([]models.Sublocation, error) {
	url := "{{baseUrl}}/sublocations/v2/inactive"
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}
	if pageNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("pageNumber=%s", pageNumber))
	}
	if pageSize != "" {
		queryParams = append(queryParams, fmt.Sprintf("pageSize=%s", pageSize))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchList[models.Sublocation](f, "Sublocations", "GetInactiveSublocations", "GET", url, nil)
}

// GetSublocationsById (Sublocations)
// GET {{baseUrl}}/sublocations/v2/{id}
//   licenseNumber (required): If specified, the Sublocation will be validated against the specified License Number. If not specified, the Sublocation will be validated against all of the User's current Facilities. Please note that if the Sublocation is not valid for the specified License Number, a 401 Unauthorized status will be returned.
func (f *Fetcher) GetSublocationsById(id string, licenseNumber string) (models.Sublocation, error) {
	url := "{{baseUrl}}/sublocations/v2/{id}"
	url = strings.ReplaceAll(url, "{"+"id"+"}", id)
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchOne[models.Sublocation](f, "Sublocations", "GetSublocationsById", "GET", url, nil)
}

// UpdateSublocations (Sublocations)
// PUT {{baseUrl}}/sublocations/v2
//   licenseNumber (required): The License Number of the Facility for which to update the list of sublocations.
func (f *Fetcher) UpdateSublocations(body []models.SublocationsRequest, licenseNumber string) (models.WriteResponse, error) {
	url := "{{baseUrl}}/sublocations/v2"
	
	queryParams := make([]string, 0)
	if licenseNumber != "" {
		queryParams = append(queryParams, fmt.Sprintf("licenseNumber=%s", licenseNumber))
	}

	if len(queryParams) > 0 {
		if strings.Contains(url, "?") {
			url += "&" + strings.Join(queryParams, "&")
		} else {
			url += "?" + strings.Join(queryParams, "&")
		}
	}
	return fetchOne[models.WriteResponse](f, "Sublocations", "UpdateSublocations", "PUT", url, body)
}
